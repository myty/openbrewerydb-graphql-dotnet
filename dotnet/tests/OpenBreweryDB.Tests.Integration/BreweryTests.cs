using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using HotChocolate.Execution;
using HotChocolate.Types.Relay;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using OpenBreweryDB.Data;
using Snapshooter.Xunit;
using Xunit;

namespace OpenBreweryDB.Tests.Integration
{
    public class BreweryTests : IClassFixture<IntegrationFixture>
    {
        private readonly IntegrationFixture _fixture;
        private readonly IIdSerializer _idSerializer;

        public BreweryTests(IntegrationFixture fixture)
        {
            _fixture = fixture;
            _fixture.ResetDatabase();

            _idSerializer = new IdSerializer();
        }

        [Fact]
        public async Task When_Query_Schema_It_ReturnResults()
        {
            var schema = await _fixture.GetSchemaAsync();

            schema
                .ToString()
                .MatchSnapshot();
        }

        [Fact]
        public async Task When_Query_Breweries_It_ReturnResults()
        {
            // Arrange
            var executor = await _fixture.GetRequestExecutorAsync();

            using var scope = _fixture.ServiceProvider.CreateScope();

            var dataContext = scope.ServiceProvider.GetService<BreweryDbContext>();
            _ = dataContext.Breweries.Add(new Data.Models.Brewery
            {
                Name = "Test",
                BreweryId = "test-id",
                Street = "123 Any St.",
                City = "My Town",
                State = "PA",
                BreweryType = "micro"
            });
            _ = dataContext.Breweries.Add(new Data.Models.Brewery
            {
                Name = "Test",
                BreweryId = "test-id-2",
                Street = "123 Any St.",
                City = "My Town",
                State = "PA",
                BreweryType = "micro"
            });

            await dataContext.SaveChangesAsync();

            // Act
            var result = await executor.ExecuteAsync(@"
                query Breweries {
                    breweries(first: 1) {
                        edges {
                            cursor
                            node {
                                name
                                brewery_id
                                id
                                street
                                city
                                state
                                brewery_type
                            }
                        }
                    }
                }
            ");

            // Assert
            result.MatchSnapshot();
        }

        [Fact]
        public async Task When_Query_Node_It_ReturnResults()
        {
            // Arrange
            var executor = await _fixture.GetRequestExecutorAsync();

            using var scope = _fixture.ServiceProvider.CreateScope();

            var dataContext = scope.ServiceProvider.GetService<BreweryDbContext>();
            var entry = dataContext.Breweries.Add(new Data.Models.Brewery
            {
                Name = "Test",
                BreweryId = "test-id",
                Street = "123 Any St.",
                City = "My Town",
                State = "PA",
                BreweryType = "micro"
            });

            await dataContext.SaveChangesAsync();

            // Act
            var queryParams = new Dictionary<string, object> {
                { "id", SerializedId(entry.Entity, (e) => e.Id) },
            };

            var result = await executor.ExecuteAsync(@"
                query Brewery($id: ID!) {
                    brewery: node(id: $id) {
                        id
                        ...BreweryBaseFields
                    }
                }

                fragment BreweryBaseFields on Brewery {
                    name
                    brewery_id
                    street
                    city
                    state
                    country
                    website_url
                    brewery_type
                }
            ", queryParams);

            // Assert
            result.MatchSnapshot();
        }

        [Fact]
        public async Task When_Query_Node_Twice_It_Returns_Two_Results()
        {
            // Arrange
            var executor = await _fixture.GetRequestExecutorAsync();

            using var scope = _fixture.ServiceProvider.CreateScope();

            var dataContext = scope.ServiceProvider.GetService<BreweryDbContext>();
            var entity1 = dataContext.Breweries.Add(new Data.Models.Brewery
            {
                Name = "Test",
                BreweryId = "test-id",
                Street = "123 Any St.",
                City = "My Town",
                State = "PA",
                BreweryType = "micro"
            });
            var entity2 = dataContext.Breweries.Add(new Data.Models.Brewery
            {
                Name = "Test",
                BreweryId = "test-id-2",
                Street = "123 Any St.",
                City = "My Town",
                State = "PA",
                BreweryType = "micro"
            });

            await dataContext.SaveChangesAsync();

            // Act
            var queryParams = new Dictionary<string, object> {
                { "id", SerializedId(entity1.Entity, (e) => e.Id) },
                { "id2", SerializedId(entity2.Entity, (e) => e.Id) }
            };

            var result = await executor.ExecuteAsync(@"
                query Brewery($id: ID!, $id2: ID!) {
                    brewery1: node(id: $id) {
                        id
                        ...BreweryBaseFields
                    }
                    brewery2: node(id: $id2) {
                        id
                        ...BreweryBaseFields
                    }
                }

                fragment BreweryBaseFields on Brewery {
                    name
                    brewery_id
                    street
                    city
                    state
                    country
                    website_url
                    brewery_type
                }
            ", queryParams);

            // Assert
            result.MatchSnapshot();
        }

        [Fact]
        public async Task When_Query_BreweryById_It_ReturnResults()
        {
            // Arrange
            var executor = await _fixture.GetRequestExecutorAsync();

            using var scope = _fixture.ServiceProvider.CreateScope();

            var dataContext = scope.ServiceProvider.GetService<BreweryDbContext>();
            _ = dataContext.Breweries.Add(new Data.Models.Brewery
            {
                Name = "Test",
                BreweryId = "test-id",
                Street = "123 Any St.",
                City = "My Town",
                State = "PA",
                BreweryType = "micro"
            });

            await dataContext.SaveChangesAsync();

            // Act
            var result = await executor.ExecuteAsync(@"
                query BreweryById($brewery_id: String!) {
                    brewery: breweryById(brewery_id: $brewery_id) {
                        id
                        ...BreweryBaseFields
                    }
                }

                fragment BreweryBaseFields on Brewery {
                    name
                    brewery_id
                    street
                    city
                    state
                    country
                    website_url
                    brewery_type
                }
            ", new Dictionary<string, object> { { "brewery_id", "test-id" } });

            // Assert
            result.MatchSnapshot();
        }

        private string SerializedId<T, V>(string typeName, T entity, Func<T, V> idValueFunc) =>
            _idSerializer.Serialize(null, typeName, idValueFunc(entity));

        private string SerializedId<T, V>(T entity, Func<T, V> idValueFunc) =>
            SerializedId(typeof(T).Name, entity, idValueFunc);
    }
}
