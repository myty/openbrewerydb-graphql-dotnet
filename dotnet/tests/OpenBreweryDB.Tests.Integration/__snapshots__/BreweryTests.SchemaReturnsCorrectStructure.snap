schema {
  query: Query
  mutation: Mutation
  subscription: Subscription
}

"The node interface is implemented by entities that have a global unique identifier."
interface Node {
  id: ID!
}

"A brewery of beer"
type Brewery implements Node {
  breweryReviews: [Review]
  "Friendly id for Brewery"
  brewery_id: String!
  "Type of Brewery"
  brewery_type: String!
  "The city of the brewery"
  city: String
  "The country of origin for the brewery"
  country: String
  createdAt: DateTime!
  id: ID!
  "Latitude portion of lat\/long coordinates"
  latitude: Decimal
  "Longitude portion of lat\/long coordinates"
  longitude: Decimal
  "Name of brewery"
  name: String!
  nearby("limit the nearby breweries to search radius, defaults to 25 milles" within: Int): [Brewery]
  "The phone number for the brewery"
  phone: String
  "The state of the brewery"
  postal_code: String
  "The state of the brewery"
  state: String
  "The street of the brewery"
  street: String
  "Tags that have been attached to the brewery"
  tag_list: [String]!
  "Date timestamp of the last time the record was updated"
  updated_at: DateTime!
  "Website address for the brewery"
  website_url: String
}

"A connection to a list of items."
type BreweryConnection {
  "A list of edges."
  edges: [BreweryEdge!]
  "A flattened list of the nodes."
  nodes: [Brewery]
  "Information to aid in pagination."
  pageInfo: PageInfo!
  totalCount: Int!
}

"An edge in a connection."
type BreweryEdge {
  "A cursor for use in pagination."
  cursor: String!
  "The item at the end of the edge."
  node: Brewery
}

type CreateBreweryPayload {
  brewery: Brewery
  "Relay Client Mutation Id"
  clientMutationId: String!
  errors: [UserError]
}

type CreateUserPayload {
  "Relay Client Mutation Id"
  clientMutationId: String!
  user: User
}

type LoginPayload {
  "Relay Client Mutation Id"
  clientMutationId: String!
  me: User
  scheme: String
  token: String
}

type Mutation {
  createBrewery(input: CreateBreweryInput): CreateBreweryPayload
  createReview(input: ReviewInput): ReviewPayload
  createUser(input: CreateUserInput): CreateUserPayload
  login(input: LoginInput): LoginPayload
  updateBrewery(input: UpdateBreweryInput): UpdateBreweryPayload
}

"Information about pagination in a connection."
type PageInfo {
  "When paginating forwards, the cursor to continue."
  endCursor: String
  "Indicates whether more edges exist following the set defined by the clients arguments."
  hasNextPage: Boolean!
  "Indicates whether more edges exist prior the set defined by the clients arguments."
  hasPreviousPage: Boolean!
  "When paginating backwards, the cursor to continue."
  startCursor: String
}

type Query {
  breweries(after: String before: String "filter by brewery id" brewery_id: String "search by city name" city: String first: PaginationAmount last: PaginationAmount "search by brewery name" name: String "general search" search: String "sort by" sort: [String] "filter by state" state: String "filter by tags" tags: [String] "filter by type" type: String): BreweryConnection
  breweryById("filter by brewery id" brewery_id: String): Brewery
  nearbyBreweries(after: String before: String first: PaginationAmount last: PaginationAmount latitude: Float! longitude: Float!): BreweryConnection
  node(id: ID!): Node
}

type Review {
  body: String!
  brewery: Brewery!
  createdOn: DateTime!
  id: Long!
  subject: String!
}

type ReviewPayload {
  "Relay Client Mutation Id"
  clientMutationId: String!
  review: Review
}

type Subscription {
  onReviewReceived: Review
}

type UpdateBreweryPayload {
  brewery: Brewery
  "Relay Client Mutation Id"
  clientMutationId: String!
  errors: [UserError]
}

type User {
  email: String!
  firstName: String!
  fullName: String
  id: Long!
  lastName: String!
}

type UserError {
  code: String
  message: String
}

input CreateBreweryInput {
  brewery_id: String!
  "Type of Brewery"
  brewery_type: String!
  "The city of the brewery"
  city: String
  "Relay Client Mutation Id"
  clientMutationId: String!
  "The country of origin for the brewery"
  country: String
  "Latitude portion of lat\/long coordinates"
  latitude: Decimal
  "Longitude portion of lat\/long coordinates"
  longitude: Decimal
  "Name of brewery"
  name: String!
  "The phone number for the brewery"
  phone: String
  "The postal code of the brewery"
  postal_code: String
  "The state of the brewery"
  state: String!
  "The street of the brewery"
  street: String
  "Tags that have been attached to the brewery"
  tag_list: [String]
  "Website address for the brewery"
  website_url: String
}

input CreateUserInput {
  "Relay Client Mutation Id"
  clientMutationId: String!
  email: String
  firstName: String
  lastName: String
  password: String
}

input LoginInput {
  "Relay Client Mutation Id"
  clientMutationId: String!
  email: String
  password: String
}

input ReviewInput {
  body: String!
  breweryId: ID!
  "Relay Client Mutation Id"
  clientMutationId: String!
  subject: String!
}

input UpdateBreweryInput {
  brewery_id: String!
  "Type of Brewery"
  brewery_type: String!
  "The city of the brewery"
  city: String
  "Relay Client Mutation Id"
  clientMutationId: String!
  "The country of origin for the brewery"
  country: String
  id: ID!
  "Latitude portion of lat\/long coordinates"
  latitude: Decimal
  "Longitude portion of lat\/long coordinates"
  longitude: Decimal
  "Name of brewery"
  name: String!
  "The phone number for the brewery"
  phone: String
  "The postal code of the brewery"
  postal_code: String
  "The state of the brewery"
  state: String!
  "The street of the brewery"
  street: String
  "Tags that have been attached to the brewery"
  tag_list: [String]
  "Website address for the brewery"
  website_url: String
}

"The `Boolean` scalar type represents `true` or `false`."
scalar Boolean

"The `DateTime` scalar represents an ISO-8601 compliant date time type."
scalar DateTime

"The built-in `Decimal` scalar type."
scalar Decimal

"The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http:\/\/en.wikipedia.org\/wiki\/IEEE_floating_point)."
scalar Float

"The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `\"4\"`) or integer (such as `4`) input value will be accepted as an ID."
scalar ID

"The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1."
scalar Int

"The `Long` scalar type represents non-fractional signed whole 64-bit numeric values. Long can represent values between -(2^63) and 2^63 - 1."
scalar Long

scalar PaginationAmount

"The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text."
scalar String
